### ISSUE TITLE
Is there a dgl-based implementation for diffpool?

### ISSUE THREAD

**Issue by jackroos** [dmlc/dgl#459]
Hi, I want to check if there is a dgl-based implementation for diffpool: https://arxiv.org/pdf/1806.08804.pdf? Thanks!

**Comment by zheng-da** [comment]
Yes. We just PRed an implementation: https://github.com/dmlc/dgl/pull/439

**Comment by yzh119** [comment]
PR #439 is in review. You're welcomed to make suggestions!

**Comment by HQ01** [comment]
Hi, thank you for your interest. PR #439 implements Diffpool in DGL, it supports batch version but still needs to be optimized. Any suggestions/help/opinion will be more than welcome!

**Comment by jackroos** [comment]
Thanks for your quick reply! LGTM. Does it reproduce the result reported in the paper? And is there an example for the usage of diffpool?

**Comment by HQ01** [comment]
TL; DR. Yes it reproduces result on ENZYME. example usage will be added in the coming days (most likely this weekend).

Yes, it can match paper result on ENZYME. For DD we probably need to improve the batch implementation first. The reason is that now DGL batches small graphs into a larger block diagonal graph (from the adjacency matrix perspective), which is not the optimal design for implementing a model like DiffPool, as DiffPool involves some tensor operations like $S^T A S$ which is hard to be formulated as message passing. With the current implmentation, DGL batched version's DiffPool will be quite slow to run on DD dataset.

**Comment by jackroos** [comment]
OK. So if batch size is small, the overhead of batched diffpool would be negligible, right?

**Comment by HQ01** [comment]
Yes. But pytorch sparse mm is still slower than batching adjacencies matrices as a tensor. I will profile it in the coming days. Notice that non-batched version on DD converges slow/doesn't converge.

**Comment by jackroos** [comment]
Thank you so much again. Look forward for your final optimized version.

**Comment by HQ01** [comment]
Actually, I wouldn't say the batch/unbatch operation is the major overhead; the main issue with current implementation is that sparse mm is slower than tensorized masked mm. Batching graphs into block diagonal batched graph further exacerbates this issue.  We will address this issue.

**Comment by HQ01** [comment]
No problem. Stay tuned! :)

**Comment by maclandrol** [comment]
@HQ01 Any update on this ? 

Also from the code, it appears that you are not including edge weight (which is, from what I have seen from the DGL documentation,  currently impossible given the nature of the adjacency matrix). 

Therefore, after each pooling layer, the new coarsened graph will always be fully connected. From my personal experience, the soft-assignment matrix of Diffpool is not sparse enough even with the regularization,  thus entries are rarely at 0.

**Comment by jermainewang** [comment]
> Also from the code, it appears that you are not including edge weight (which is, from what I have seen from the DGL documentation, currently impossible given the nature of the adjacency matrix).

You could put edge weights as edge features through `g.edata`.

**Comment by HQ01** [comment]
We are working on optimizing DGL's BatchedGraph API to optimize for Diffpool (and other potential usages). @VoVAllen may give you more details as he's taking the lead on this effort. 

As for "edge weight" -- I guess you are referring to the entry values in the adjacency matrix. They are indeed not included in the graph exactly because of your correct observation that after the first pooling layer, the coarsened graph is dense. Instead, after the first pooling layer, we simply store the new "graph" as an adjacency matrix. Another thing to point out is that even though each adjacency matrix is dense, once you batch multiple graphs and compute the new "coarsened" batched graph, the "coarsened" batched graph's adjacency matrix will be dense block diagonal instead of fully dense.

**Comment by crazysal** [comment]
https://discuss.dgl.ai/t/diffpool-with-nn-layers/690

#tldr. 
Is the example script provided the best way to implement difPool and it requires a custom implementation of the underlying graph nn or can the dgl.nn modules take updated adgacency matrices ??
